Update nalgebra to 0.12

This will require fixing `crates/types/src/vector3.rs`:

```rust
    #[test]
    fn test_typical_normalize_use_case() {
        let position: Vector3 = Vector3::new(100.into(), 100.into(), 1.into());
        let velocity: Vector3 = Vector3::new(1.into(), 2.into(), 0.5.into());
        let time: Fixed = 0.0166666667.into();

        let new_position = position + velocity * time;
        let delta = new_position - position;
        let direction = delta.normalized();
        let direction_length = direction.length();

        assert_eq!("(100.0166667, 100.0333333, 1.0083333)".to_string(),
                   format!("{}", new_position));
        assert_eq!("(0.0166667, 0.0333333, 0.0083333)".to_string(),
                   format!("{}", delta));
        assert_eq!("(0.4364425, 0.8728850, 0.2182212)".to_string(),
                   format!("{}", direction));
        assert_eq!(1, direction_length.to_i32().unwrap());
    }
```

```
  --> src/vector3.rs:78:34
   |
78 |                    format!("{}", new_position));
   |                                  ^^^^^^^^^^^^ the trait `alga::general::real::Real` is not implemented for `fixed::Fixed`
   |
   = note: required because of the requirements on the impl of `std::fmt::Display` for `nalgebra::Matrix<fixed::Fixed, nalgebra::U3, nalgebra::U1, nalgebra::MatrixArray<fixed::Fixed, nalgebra::U3, nalgebr
a::U1>>`
   = note: required by `std::fmt::Display::fmt`

error[E0277]: the trait bound `fixed::Fixed: alga::general::real::Real` is not satisfied
  --> src/vector3.rs:80:34
   |
80 |                    format!("{}", delta));
   |                                  ^^^^^ the trait `alga::general::real::Real` is not implemented for `fixed::Fixed`
   |
   = note: required because of the requirements on the impl of `std::fmt::Display` for `nalgebra::Matrix<fixed::Fixed, nalgebra::U3, nalgebra::U1, nalgebra::MatrixArray<fixed::Fixed, nalgebra::U3, nalgebr
a::U1>>`
   = note: required by `std::fmt::Display::fmt`

error[E0277]: the trait bound `fixed::Fixed: alga::general::real::Real` is not satisfied
  --> src/vector3.rs:82:34
   |
82 |                    format!("{}", direction));
   |                                  ^^^^^^^^^ the trait `alga::general::real::Real` is not implemented for `fixed::Fixed`
   |
   = note: required because of the requirements on the impl of `std::fmt::Display` for `nalgebra::Matrix<fixed::Fixed, nalgebra::U3, nalgebra::U1, nalgebra::MatrixArray<fixed::Fixed, nalgebra::U3, nalgebr
a::U1>>`
   = note: required by `std::fmt::Display::fmt`
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,7 +19,7 @@ opt-level = 1
 [dependencies]
 clap = "2.17"
 lazy_static = "0.2"
-nalgebra = "0.10"
+nalgebra = "0.12"
 num = "0.1"
 specs = "0.7"
 time = "0.1"

--- a/crates/media/Cargo.toml
+++ b/crates/media/Cargo.toml
@@ -10,7 +10,7 @@ authors = ["Kevin Fuller <angered.ghandi@gmail.com>"]
 
 [dependencies]
 error-chain = "0.5"
-nalgebra = "0.10"
+nalgebra = "0.12"
 
 [dependencies.sdl2]
 git = "https://github.com/AngryLawyer/rust-sdl2.git"

--- a/crates/resource/Cargo.toml
+++ b/crates/resource/Cargo.toml
@@ -10,7 +10,7 @@ authors = ["Kevin Fuller <angered.ghandi@gmail.com>"]
 
 [dependencies]
 error-chain = "0.5"
-nalgebra = "0.10"
+nalgebra = "0.12"
 chariot_drs = "0.1"
 chariot_slp = "0.1"
 chariot_palette = "0.1"

--- a/crates/types/Cargo.toml
+++ b/crates/types/Cargo.toml
@@ -9,7 +9,7 @@ keywords = ["aoe", "age", "empires", "genie", "engine"]
 authors = ["Kevin Fuller <angered.ghandi@gmail.com>"]
 
 [dependencies]
-nalgebra = "0.10"
+nalgebra = "0.12"
 num = "0.1"
 
 [dependencies.sdl2]

--- a/crates/types/src/vector3.rs
+++ b/crates/types/src/vector3.rs
@@ -63,6 +63,14 @@ mod test {
     use super::*;
     use super::super::Fixed;
 
+    fn vec3_string(vec3: &Vector3) -> String {
+        format!("({}, {}, {})",
+            vec3.x.to_f32().unwrap(),
+            vec3.y.to_f32().unwrap(),
+            vec3.z.to_f32().unwrap()
+        )
+    }
+
     #[test]
     fn test_typical_normalize_use_case() {
         let position: Vector3 = Vector3::new(100.into(), 100.into(), 1.into());
@@ -74,12 +82,9 @@ mod test {
         let direction = delta.normalized();
         let direction_length = direction.length();
 
-        assert_eq!("(100.0166667, 100.0333333, 1.0083333)".to_string(),
-                   format!("{}", new_position));
-        assert_eq!("(0.0166667, 0.0333333, 0.0083333)".to_string(),
-                   format!("{}", delta));
-        assert_eq!("(0.4364425, 0.8728850, 0.2182212)".to_string(),
-                   format!("{}", direction));
+        assert_eq!("(100.01667, 100.03333, 1.0083333)", &vec3_string(&new_position));
+        assert_eq!("(0.01666665, 0.0333333, 0.008333325)", &vec3_string(&delta));
+        assert_eq!("(0.4364425, 0.87288505, 0.21822125)", &vec3_string(&direction));
         assert_eq!(1, direction_length.to_i32().unwrap());
     }
 

--- a/src/game/state/scenario_game_state.rs
+++ b/src/game/state/scenario_game_state.rs
@@ -23,7 +23,7 @@ use ecs;
 use ecs::resource::{KeyboardKeyStates, MouseState, RenderCommands, Viewport};
 use game::{Game, GameState};
 use media::MediaRef;
-use nalgebra::{Cast, Vector2};
+use nalgebra::{Vector2, convert};
 use resource::ShapeManagerRef;
 use scn;
 use types::Fixed;
@@ -45,7 +45,7 @@ impl ScenarioGameState {
 
     fn update_viewport(&mut self, lerp: Fixed) {
         let viewport = self.planner.mut_world().read_resource::<Viewport>();
-        let top_left: Vector2<i32> = Cast::from(viewport.lerped_top_left(lerp));
+        let top_left: Vector2<i32> = convert(viewport.lerped_top_left(lerp));
         self.media.borrow_mut().renderer().set_camera_position(&top_left);
     }
 

